# ESG Risk Intelligence Platform
## Data Cleaning & Preprocessing Pipeline

This notebook focuses on preparing environmental and sustainability data for ESG risk analysis.

Objectives:
- Understand dataset structure
- Handle missing values
- Select business-relevant features
- Prepare analytics-ready dataset

## Step 1 — Import Required Libraries

The project uses Python libraries for:
- data manipulation
- numerical operations
- environmental data preprocessing

In [52]:
import pandas as pd
import numpy as np

## Step 2 — Load Raw Dataset

The raw ESG dataset is loaded from the local project directory.

Dataset Source:
Our World in Data CO2 & Greenhouse Gas Dataset

In [53]:
df = pd.read_csv('owid_co2_data.csv')

## Step 3 — Initial Dataset Inspection

Initial inspection helps understand:
- dataset size
- column structure
- overall data quality

In [54]:
df.head()

,country,year,iso_code,population,gdp,cement_co2,cement_co2_per_capita,co2,co2_growth_abs,co2_growth_prct,...,share_global_other_co2,share_of_temperature_change_from_ghg,temperature_change_from_ch4,temperature_change_from_co2,temperature_change_from_ghg,temperature_change_from_n2o,total_ghg,total_ghg_excluding_lucf,trade_co2,trade_co2_share
0,Afghanistan,1750,AFG,2802560.0,NaN,0.0,0.0,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Afghanistan,1751,AFG,NaN,NaN,0.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Afghanistan,1752,AFG,NaN,NaN,0.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Afghanistan,1753,AFG,NaN,NaN,0.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Afghanistan,1754,AFG,NaN,NaN,0.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [55]:
df.shape

(50411, 79)

## Step 4 — Dataset Information

This step identifies:
- data types
- missing values
- memory usage
- numerical and categorical features

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50411 entries, 0 to 50410
Data columns (total 79 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   country                                    50411 non-null  object 
 1   year                                       50411 non-null  int64  
 2   iso_code                                   42480 non-null  object 
 3   population                                 41167 non-null  float64
 4   gdp                                        15251 non-null  float64
 5   cement_co2                                 29173 non-null  float64
 6   cement_co2_per_capita                      25648 non-null  float64
 7   co2                                        29384 non-null  float64
 8   co2_growth_abs                             27216 non-null  float64
 9   co2_growth_prct                            26239 non-null  float64
 10  co2_including_luc     

## Step 5 — Missing Value Analysis

The dataset contains multiple environmental indicators with varying levels of missing data.

A structured missing-value analysis is performed before cleaning decisions are made.

In [57]:
missing = df.isnull().sum().sort_values(ascending=False)

missing_percentage = (
    df.isnull().mean() * 100
).sort_values(ascending=False)

missing_df = pd.DataFrame({
    'Missing Values': missing,
    'Missing Percentage': missing_percentage
})

missing_df.head(20)

,Missing Values,Missing Percentage
share_global_cumulative_other_co2,48241,95.695384
share_global_other_co2,48241,95.695384
other_co2_per_capita,47752,94.725358
other_industry_co2,47157,93.545060
cumulative_other_co2,47157,93.545060
consumption_co2_per_gdp,45963,91.176529
consumption_co2_per_capita,45768,90.789709
trade_co2,45699,90.652834
trade_co2_share,45699,90.652834
consumption_co2,45358,89.976394


## Step 6 — Business-Relevant Feature Selection

Only ESG-relevant columns were selected to improve analytical focus and reduce unnecessary complexity.

Selected features represent:
- emissions
- energy usage
- economic indicators
- climate impact

In [58]:
important_columns = [
    'country',
    'year',
    'population',
    'gdp',
    'co2',
    'co2_per_capita',
    'coal_co2',
    'oil_co2',
    'gas_co2',
    'methane',
    'energy_per_capita',
    'share_global_co2',
    'temperature_change_from_ghg',
    'total_ghg'
]

esg_df = df[important_columns].copy()

## Step 7 — Create ESG Working Dataset

A separate working dataframe was created to preserve raw data integrity.

In [59]:
esg_df.head()

,country,year,population,gdp,co2,co2_per_capita,coal_co2,oil_co2,gas_co2,methane,energy_per_capita,share_global_co2,temperature_change_from_ghg,total_ghg
0,Afghanistan,1750,2802560.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Afghanistan,1751,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Afghanistan,1752,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Afghanistan,1753,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Afghanistan,1754,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [60]:
esg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50411 entries, 0 to 50410
Data columns (total 14 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   country                      50411 non-null  object 
 1   year                         50411 non-null  int64  
 2   population                   41167 non-null  float64
 3   gdp                          15251 non-null  float64
 4   co2                          29384 non-null  float64
 5   co2_per_capita               26509 non-null  float64
 6   coal_co2                     21925 non-null  float64
 7   oil_co2                      25459 non-null  float64
 8   gas_co2                      18147 non-null  float64
 9   methane                      38150 non-null  float64
 10  energy_per_capita            10459 non-null  float64
 11  share_global_co2             27563 non-null  float64
 12  temperature_change_from_ghg  41238 non-null  float64
 13  total_ghg       

## Step 8 — Analyze Missing Values in ESG Dataset

After selecting ESG-relevant features, missing values were re-evaluated to determine appropriate cleaning strategies for each variable.

The objective is to:
- preserve analytical quality
- minimize unnecessary data loss
- maintain long-term environmental trends

In [61]:
esg_df.isnull().sum()

country                            0
year                               0
population                      9244
gdp                            35160
co2                            21027
co2_per_capita                 23902
coal_co2                       28486
oil_co2                        24952
gas_co2                        32264
methane                        12261
energy_per_capita              39952
share_global_co2               22848
temperature_change_from_ghg     9173
total_ghg                      12261
dtype: int64

## Step 9 — Filter Modern Analytical Period

The original dataset contained historical records dating back to 1750.

However, early records contained limited environmental and economic reporting.

To improve analytical reliability and consistency, the dataset was filtered to modern reporting years where ESG indicators are more complete and meaningful.

In [62]:
esg_df = esg_df[
    esg_df['year'] >= 1950
]

## Step 10 — Validate Dataset After Year Filtering

The dataset was rechecked after filtering to evaluate improvements in data completeness and analytical quality.

In [63]:
esg_df.shape


(18984, 14)

In [64]:
esg_df.isnull().sum()

country                           0
year                              0
population                     1809
gdp                            7512
co2                            1437
co2_per_capita                 2491
coal_co2                       7027
oil_co2                        2365
gas_co2                        9554
methane                        2634
energy_per_capita              8525
share_global_co2               2258
temperature_change_from_ghg    1209
total_ghg                      2634
dtype: int64

## Step 11 — Remove Critical Missing Records

Rows missing essential environmental indicators such as CO2 emissions and greenhouse gas metrics were removed because they cannot support meaningful ESG risk analysis.

In [65]:
esg_df = esg_df.dropna(
    subset=['co2', 'total_ghg']
)

## Step 12 — Reassess Remaining Missing Values

After removing critical missing records, remaining null values were analyzed to determine feature-specific imputation strategies.

In [66]:
esg_df.isnull().sum()

country                           0
year                              0
population                      375
gdp                            4372
co2                               0
co2_per_capita                    0
coal_co2                       4243
oil_co2                          54
gas_co2                        6517
methane                           0
energy_per_capita              5527
share_global_co2                  0
temperature_change_from_ghg       0
total_ghg                         0
dtype: int64

## Step 13 — Handle Missing Economic Indicators

Median imputation was applied to economic variables because it provides stability against extreme outliers commonly found in macroeconomic data.

In [67]:
esg_df['gdp'] = esg_df['gdp'].fillna(
    esg_df['gdp'].median()
)

esg_df['population'] = esg_df['population'].fillna(
    esg_df['population'].median()
)

## Step 14 — Interpolate Environmental Metrics

Environmental indicators typically evolve gradually over time.

Country-wise interpolation was applied to preserve yearly environmental trends while minimizing information loss.

In [68]:
esg_df = esg_df.sort_values(
    ['country', 'year']
)

environmental_cols = [
    'methane',
    'energy_per_capita',
    'co2_per_capita'
]

esg_df[environmental_cols] = (
    esg_df.groupby('country')[
        environmental_cols
    ].transform(
        lambda x: x.interpolate()
    )
)

## Step 15 — Final Missing Value Validation

A final validation step was performed to identify any remaining missing observations before preparing the dataset for downstream analytics and SQL integration.

In [69]:
esg_df.isnull().sum()

country                           0
year                              0
population                        0
gdp                               0
co2                               0
co2_per_capita                    0
coal_co2                       4243
oil_co2                          54
gas_co2                        6517
methane                           0
energy_per_capita              5409
share_global_co2                  0
temperature_change_from_ghg       0
total_ghg                         0
dtype: int64

## Step 16 — Handle Residual Environmental Missing Values

After interpolation and economic imputation, several environmental indicators still contained missing observations.

Instead of removing all incomplete rows, feature-specific domain logic was applied to preserve analytical coverage and minimize unnecessary data loss.

Environmental emission columns were treated separately because missing values may represent:
- negligible industrial activity
- unavailable reporting
- low fossil-fuel dependency

This approach improves dataset retention while maintaining analytical consistency.

### Emission Column Treatment

Remaining missing values in fossil-fuel emission columns were replaced with zero values.

This decision was made because many countries and historical records reported negligible or unavailable industrial fuel emissions.

In [70]:
list(esg_df.columns)

['country',
 'year',
 'population',
 'gdp',
 'co2',
 'co2_per_capita',
 'coal_co2',
 'oil_co2',
 'gas_co2',
 'methane',
 'energy_per_capita',
 'share_global_co2',
 'temperature_change_from_ghg',
 'total_ghg']

In [71]:
emission_cols = [
    'coal_co2',
    'oil_co2',
    'gas_co2'
]

esg_df[emission_cols] = (
    esg_df[emission_cols]
    .fillna(0)
)

### Energy Consumption Imputation

Country-wise forward-fill and backward-fill techniques were applied to energy consumption indicators to preserve temporal continuity across yearly observations.

In [72]:
esg_df['energy_per_capita'] = (
    esg_df.groupby('country')[
        'energy_per_capita'
    ]
    .transform(
        lambda x: x.ffill().bfill()
    )
)

## Step 17 — Validate Residual Missing Values

The dataset was revalidated after feature-specific environmental imputations to ensure improved completeness and analytical consistency.

In [73]:
esg_df.isnull().sum()

country                          0
year                             0
population                       0
gdp                              0
co2                              0
co2_per_capita                   0
coal_co2                         0
oil_co2                          0
gas_co2                          0
methane                          0
energy_per_capita              546
share_global_co2                 0
temperature_change_from_ghg      0
total_ghg                        0
dtype: int64

## Step 18 — Remove Residual Incomplete Records

A minimal number of residual incomplete observations remained after domain-specific imputation.

These records were removed to finalize a fully analysis-ready ESG dataset.

A small number of residual missing observations remained after feature-specific imputations.

These records primarily belonged to countries with insufficient long-term energy reporting and represented a minor portion of the analytical dataset.

The remaining incomplete observations were removed to finalize a fully analysis-ready ESG dataset.

In [74]:
esg_df = esg_df.dropna()

In [75]:
esg_df.shape

(15045, 14)

In [76]:
esg_df.head()

,country,year,population,gdp,co2,co2_per_capita,coal_co2,oil_co2,gas_co2,methane,energy_per_capita,share_global_co2,temperature_change_from_ghg,total_ghg
200,Afghanistan,1950,7776180.0,9.421400e+09,0.084,0.011,0.021,0.063,0.0,7.879,481.209,0.001,0.001,19.869
201,Afghanistan,1951,7879343.0,9.692280e+09,0.092,0.012,0.026,0.066,0.0,7.973,481.209,0.001,0.001,21.069
202,Afghanistan,1952,7987784.0,1.001732e+10,0.092,0.011,0.032,0.060,0.0,8.073,481.209,0.001,0.001,22.094
203,Afghanistan,1953,8096703.0,1.063052e+10,0.106,0.013,0.038,0.068,0.0,8.186,481.209,0.002,0.001,23.256
204,Afghanistan,1954,8207954.0,1.086636e+10,0.106,0.013,0.043,0.064,0.0,8.305,481.209,0.002,0.001,24.251


## Step 19 — Duplicate Record Validation

Duplicate validation was performed to ensure dataset integrity and avoid analytical distortion during ESG trend analysis.

Maintaining unique country-year observations is essential for:
- accurate time-series analysis
- sustainability trend evaluation
- ESG risk modeling

In [84]:
duplicate_count = esg_df.duplicated().sum()

print(
    f"Total Duplicate Records: {duplicate_count}"
)

Total Duplicate Records: 0


## Step 20— Final Dataset Summary

The ESG dataset was successfully cleaned and transformed into an analysis-ready format.

Final preprocessing outcomes:
- reduced feature complexity
- improved data completeness
- preserved environmental trend continuity
- minimized unnecessary data loss
- prepared structured ESG indicators for downstream analytics and dashboard development

In [79]:
esg_df.shape

(15045, 14)

In [80]:
esg_df.describe()

,year,population,gdp,co2,co2_per_capita,coal_co2,oil_co2,gas_co2,methane,energy_per_capita,share_global_co2,temperature_change_from_ghg,total_ghg
count,15045.000000,1.504500e+04,1.504500e+04,15045.000000,15045.00000,15045.000000,15045.000000,15045.000000,15045.000000,15045.000000,15045.000000,15045.000000,15045.000000
mean,1987.833500,1.045273e+08,3.236263e+11,438.039107,4.28236,180.973227,159.638448,74.655780,139.255016,21110.319793,2.037629,0.019753,743.969183
std,21.482672,5.034136e+08,3.105837e+12,2264.417120,7.31930,977.060114,814.888942,420.245355,618.042719,29951.326384,9.730536,0.093202,3451.431436
min,1950.000000,1.776000e+03,5.543200e+07,0.000000,0.00000,0.000000,0.000000,0.000000,0.001000,0.000000,0.000000,-0.001000,-9.219000
25%,1969.000000,1.561183e+06,1.544460e+10,0.879000,0.39800,0.000000,0.629000,0.000000,1.936000,2074.009000,0.005000,0.000000,9.039000
50%,1988.000000,6.033296e+06,3.452829e+10,7.722000,1.75000,0.256000,3.935000,0.007000,9.253000,9163.633000,0.039000,0.001000,42.083000
75%,2006.000000,2.228490e+07,9.091082e+10,59.622000,5.93100,10.874000,26.810000,7.288000,34.064000,28982.496000,0.279000,0.004000,140.441000
max,2024.000000,8.161973e+09,1.301126e+14,38598.578000,364.79100,15805.254000,12470.596000,8009.828000,9498.914000,318559.688000,100.000000,1.678000,54433.398000


In [81]:
esg_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15045 entries, 200 to 50410
Data columns (total 14 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   country                      15045 non-null  object 
 1   year                         15045 non-null  int64  
 2   population                   15045 non-null  float64
 3   gdp                          15045 non-null  float64
 4   co2                          15045 non-null  float64
 5   co2_per_capita               15045 non-null  float64
 6   coal_co2                     15045 non-null  float64
 7   oil_co2                      15045 non-null  float64
 8   gas_co2                      15045 non-null  float64
 9   methane                      15045 non-null  float64
 10  energy_per_capita            15045 non-null  float64
 11  share_global_co2             15045 non-null  float64
 12  temperature_change_from_ghg  15045 non-null  float64
 13  total_ghg          

In [82]:
esg_df['share_global_co2'].describe()

count    15045.000000
mean         2.037629
std          9.730536
min          0.000000
25%          0.005000
50%          0.039000
75%          0.279000
max        100.000000
Name: share_global_co2, dtype: float64

## Step 21 — Export Final ESG Dataset

The final ESG dataset was exported for downstream:
- SQL analysis
- ESG risk modeling
- dashboard development
- sustainability intelligence reporting

In [85]:
esg_df.to_csv(
    'final_esg_dataset.csv',
    index=False
)